### PR TITLE
Support fp16 in synchronized batchnorm

### LINKED ISSUE #12257: SyncBatchNorm NOT Supporting Mix Precision Training for Float16
Current SyncBatchNorm does not consider the mix precision training for float16. Standard BatchNorm allows using float32 params with float16 data.

http://mxnet.incubator.apache.org/versions/master/api/python/gluon/contrib.html?highlight=syncbatchnorm#mxnet.gluon.contrib.nn.SyncBatchNorm

### PR DESCRIPTION
**Pull request by haojin2**
## Description ##
As title, fix #12257.

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Support fp16 `data` input in forward pass
- [x] Support fp16 `data` input in backward pass
- [x] Test coverage for both passes

## Comments ##
All other inputs and outputs remain fp32 except the `data`.

### CODE REVIEW

File: src/operator/contrib/sync_batch_norm-inl.h [hunk 1]
```diff
@@ -271,69 +273,102 @@ class SyncBatchNorm : public Operator {
...
+          Shape1(data_size * 2), s);
+      }
+      if (in_data[syncbatchnorm::kData].ndim() == 2) {
+        Shape<4> dshape = Shape4(in_data[syncbatchnorm::kData].shape_[0],
+                                in_data[syncbatchnorm::kData].shape_[1], 1, 1);
+        if (std::is_same<DType, real_t>::value) {
+          data = in_data[syncbatchnorm::kData].get_with_shape<xpu, 4, real_t>(dshape, s);
+          out = out_data[syncbatchnorm::kOut].get_with_shape<xpu, 4, real_t>(dshape, s);
+        } else {
+          data = Tensor<xpu, 4>(workspace.dptr_, dshape, s);
```
Comment (apeforest) [inline src/operator/contrib/sync_batch_norm-inl.h]: I think we should write a function such as getData() to handle the different ways to get data between half_t and real_t case. So many repetition if branches in this code makes it hard to read and bug prone.

File: tests/python/gpu/test_gluon_gpu.py [hunk 2]
```diff
@@ -193,10 +193,12 @@ def get_num_devices():
     if get_num_devices() < 2:
         return
     ndev = 2
+    dtypes = [np.float16, np.float32]
```
Comment (apeforest) [inline tests/python/gpu/test_gluon_gpu.py]: It's better to also test float64 and assert the exception here.

File: src/operator/contrib/sync_batch_norm-inl.h [hunk 3]
```diff
@@ -271,69 +273,102 @@ class SyncBatchNorm : public Operator {
...
+      const bool is_double = std::is_same<DType, double>::value;
+      CHECK_EQ(is_double, false)
+        << "Synchronized BatchNorm does not support double-precision floating number yet...";
+      const real_t scale = static_cast<real_t>(in_data[syncbatchnorm::kData].shape_[1]) /
+        static_cast<real_t>(in_data[syncbatchnorm::kData].shape_.Size());
+      const size_t data_size = in_data[syncbatchnorm::kData].Size();
+      Tensor<xpu, 4> data;
+      Tensor<xpu, 4> out;
+      Tensor<xpu, 1> workspace;
+      if (!std::is_same<DType, real_t>::value) {
```
[the remark this answers is not in the document]
Comment (haojin2) [reply src/operator/contrib/sync_batch_norm-inl.h]: this is for an easier support for float64 in the future.

File: src/operator/contrib/sync_batch_norm-inl.h [hunk 4]
```diff
@@ -345,109 +380,161 @@ class SyncBatchNorm : public Operator {
...
+      }
+      if (!std::is_same<DType, real_t>::value) {
+        total_workspace_size += 3 * data_size;
+      }
+
+      workspace = ctx.requested[syncbatchnorm::kTempSpace].get_space<xpu, 1>(
+                    mshadow::Shape1(total_workspace_size), s);
+
+      const real_t scale = static_cast<real_t>(out_grad[syncbatchnorm::kOut].shape_[1]) /
+        static_cast<real_t>(out_grad[syncbatchnorm::kOut].shape_.Size());
```
[the remark this answers is not in the document]
Comment (haojin2) [reply src/operator/contrib/sync_batch_norm-inl.h]: you actually need, both nominator and denominator are integer types.

File: src/operator/contrib/sync_batch_norm-inl.h [hunk 5]
```diff
@@ -271,69 +273,102 @@ class SyncBatchNorm : public Operator {
...
-                                          data.shape_) * data +
-             broadcast<1>(bias - (slope * moving_mean) /
-                          F<mshadow_op::square_root>(moving_var + param_.eps), data.shape_));
-    }
+    MSHADOW_TYPE_SWITCH(in_data[syncbatchnorm::kData].type_flag_, DType, {
+      const bool is_double = std::is_same<DType, double>::value;
+      CHECK_EQ(is_double, false)
+        << "Synchronized BatchNorm does not support double-precision floating number yet...";
+      const real_t scale = static_cast<real_t>(in_data[syncbatchnorm::kData].shape_[1]) /
+        static_cast<real_t>(in_data[syncbatchnorm::kData].shape_.Size());
```
[the remark this answers is not in the document]
Comment (haojin2) [reply src/operator/contrib/sync_batch_norm-inl.h]: you actually need, both nominator and denominator are integer types.

File: src/operator/contrib/sync_batch_norm-inl.h (hunk 1 above)
Comment (haojin2) [reply src/operator/contrib/sync_batch_norm-inl.h]: If you think of it carefully you'll see that such a function will have quite a few arguments and still very hard to debug and will introduce extra overhead.

File: src/operator/contrib/sync_batch_norm-inl.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (apeforest) [reply src/operator/contrib/sync_batch_norm-inl.h]: But you are using division here.

File: src/operator/contrib/sync_batch_norm-inl.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (apeforest) [reply src/operator/contrib/sync_batch_norm-inl.h]: But your implementation is only for float16

File: src/operator/contrib/sync_batch_norm-inl.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (apeforest) [reply src/operator/contrib/sync_batch_norm-inl.h]: division will automatically cast the result to float if one of them is float

File: src/operator/contrib/sync_batch_norm-inl.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (haojin2) [reply src/operator/contrib/sync_batch_norm-inl.h]: It's on the road map so I'm planning for the future here.

File: src/operator/contrib/sync_batch_norm-inl.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (haojin2) [reply src/operator/contrib/sync_batch_norm-inl.h]: output from this simple program:
```c++
#include <iostream>

using namespace std;

int main() {
  cout << static_cast<float>(1/2) << endl;
  cout << static_cast<float>(1) / static_cast<float>(2) << endl;
  cout << static_cast<float>(3/2) << endl;
  cout << static_cast<float>(3) / static_cast<float>(2) << endl;
  return 0;
}
```
is
```
0
0.5
0
0.5
1
1.5
```
So you do need it

File: src/operator/contrib/sync_batch_norm-inl.h (hunk 1 above)
Comment (apeforest) [reply src/operator/contrib/sync_batch_norm-inl.h]: I think there are only a few arguments: (TBlob &, DType, Stream &). What is the extra overhead do you mean? On the other hand, I can see this same logic block being used at least 4 times in this code.

File: src/operator/contrib/sync_batch_norm-inl.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (haojin2) [reply src/operator/contrib/sync_batch_norm-inl.h]: Ok I see what you mean, but I'm simply copying the original code here and I would prefer not touching anything not necessary to my addition.

[the remark this answers is not in the document]
Comment (apeforest) [reply src/operator/contrib/sync_batch_norm-inl.h]: But the code in this file is actually the following case:

```
#include <iostream>

using namespace std;

int main() {
  cout << static_cast<float>(1)/2 << endl;
  cout << static_cast<float>(1)/static_cast<float>(2) << endl;
  return 0;
}

output is:
0.5
0.5
```

File: src/operator/contrib/sync_batch_norm-inl.h (hunk 1 above)
Comment (haojin2) [reply src/operator/contrib/sync_batch_norm-inl.h]: no you actually need more than that including `in_data[syncbatchnorm::kData]`, `out_data[syncbatchnorm::kOut]`, `s`, `workspace.dptr_`, `dshape` and some of them need to be null/nothing in some cases, so I don't it's a very suitable case for a "common" helper function.

Comment (apeforest) [reply src/operator/contrib/sync_batch_norm-inl.h]: Both in_data[syncbatchnorm::kData] and out_data[syncbatchnorm::kOut] are TBlob, you only need to write a common function to process both of them so only one TBlob argument is needed. As for workspace, I think you can wrap that inside the function as well with some proper refactoring.

**Comment by haojin2** [comment]
Closing due to the implementation is pending verification on actual training workloads.

File: src/operator/contrib/sync_batch_norm-inl.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (apeforest) [reply src/operator/contrib/sync_batch_norm-inl.h]: But the logic here does not work for float64. If you want to support float64, it is probably better to have a specific logic block. For now, this logic literally means "if DType is not float32, do this..." which is incorrect.

File: src/operator/contrib/sync_batch_norm-inl.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (apeforest) [reply src/operator/contrib/sync_batch_norm-inl.h]: Since you are inserting logic to the original code block, you are already modifying the code, isn't it  :-)
